### PR TITLE
Support Xcode betas with xcode-locator

### PR DESCRIPTION
**Pull request by keith**
Previously if you had 2 builds of the same version of Xcode, for example
10.2.0 beta 1 and 10.2.0 beta 2, there was no way to disambiguate them.
This change appends the `ProductBuildVersion` of each Xcode version to
allow you to specify that if desired.

### CODE REVIEW

**Comment by keith** [comment]
Here is the example output in the previous case:

```
% bazel-bin/tools/osx/xcode-locator 2>/dev/null
{
        "10": "/Applications/Xcode-10.1.0.app/Contents/Developer",
        "10.2.0": "/Applications/Xcode-10.2.0b1.app/Contents/Developer",
        "10.1.0": "/Applications/Xcode-10.1.0.app/Contents/Developer",
        "10.1": "/Applications/Xcode-10.1.0.app/Contents/Developer",
        "10.2": "/Applications/Xcode-10.2.0b2.app/Contents/Developer",
}
% bazel-bin/tools/osx/xcode-locator -v 2>/dev/null
10.2.0:10.2,10.2.0:/Applications/Xcode-10.2.0b1.app/Contents/Developer
10.1.0:10.1.0,10.1,10:/Applications/Xcode-10.1.0.app/Contents/Developer
% bazel-bin/tools/osx/xcode-locator 10.2.0 2>/dev/null
/Applications/Xcode-10.2.0b1.app/Contents/Developer
```

As you can see there isn't a single version you can pass to make sure you get beta 1 or beta 2. I also assume this differs based on what the OS returns from `LSCopyApplicationURLsForBundleIdentifier`.

Here's the new output in this case:

```
% bazel-bin/tools/osx/xcode-locator 2>/dev/null
{
        "10": "/Applications/Xcode-10.1.0.app/Contents/Developer",
        "10.1": "/Applications/Xcode-10.1.0.app/Contents/Developer",
        "10.1.0": "/Applications/Xcode-10.1.0.app/Contents/Developer",
        "10.2.0.14490.95": "/Applications/Xcode-10.2.0b2.app/Contents/Developer",
        "10.2.0.14490": "/Applications/Xcode-10.2.0b2.app/Contents/Developer",
        "10.2.0": "/Applications/Xcode-10.2.0b2.app/Contents/Developer",
        "10.1.0.14460": "/Applications/Xcode-10.1.0.app/Contents/Developer",
        "10.1.0.14460.46": "/Applications/Xcode-10.1.0.app/Contents/Developer",
        "10.2.0.14490.87.2": "/Applications/Xcode-10.2.0b1.app/Contents/Developer",
        "10.2.0.14490.87": "/Applications/Xcode-10.2.0b1.app/Contents/Developer",
        "10.2.0.14490.95.3": "/Applications/Xcode-10.2.0b2.app/Contents/Developer",
        "10.2": "/Applications/Xcode-10.2.0b2.app/Contents/Developer",
}
% bazel-bin/tools/osx/xcode-locator -v 2>/dev/null
10.2.0:10.2.0.14490.87,10.2.0,10.2.0.14490.95.3,10.2,10.2.0.14490.87.2,10.2.0.14490.95,10.2.0.14490:/Applications/Xcode-10.2.0b2.app/Contents/Developer
10.1.0:10.1.0,10,10.1,10.1.0.14460,10.1.0.14460.46:/Applications/Xcode-10.1.0.app/Contents/Developer
% bazel-bin/tools/osx/xcode-locator 10.2.0 2>/dev/null
/Applications/Xcode-10.2.0b2.app/Contents/Developer
% bazel-bin/tools/osx/xcode-locator 10.2.0.14490.95.3 2>/dev/null
/Applications/Xcode-10.2.0b2.app/Contents/Developer
% bazel-bin/tools/osx/xcode-locator 10.2.0.14490.87 2>/dev/null
/Applications/Xcode-10.2.0b1.app/Contents/Developer
```

There are some downsides to this representation. For one it's still ambiguous when using `xcode-locator -v`. This shouldn't be a big problem since this feature won't be used by everyone, and you can see the longer versions if you don't pass `-v`. The other downside is this version isn't particularly user facing. I would have preferred to use the build number that Xcode shows in the "Welcome to Xcode" UI but there are some problems with that. First off that version isn't in the Info.plist, it's in Contents/version.plist, meaning we have to parse a second file to get it. The bigger problem is that DottedVersion doesn't like having letters in sections and crashes on certain patterns of that. I figured this option was better than working around those downsides because it will be a lesser used feature anyways.

**Comment by sergiocampama** [comment]
`DottedVersion` requires each component to follow the `\d+[a-Z]*\d*` regexp. What if when the component starts with a letter, you append a 0 in the beginning? I think that should be semver equivalent.

I like this change, it improves xcode locator's functionality quite a lot.

looping in @aragos and @c-parsons for input from the bazel POV.

**Comment by aragos** [comment]
Right, `DottedVersion` actually follows Apple's naming scheme - if you look at the directory names you have your Xcodes in they work with `DottedVersion`. 

In general I'd be ok with this change but shorter/more descriptive version names would be nice. I don't think parsing a second file would be a problem (this is cached, right?) but @c-parsons may disagree. :)

**Comment by keith** [comment]
Do we know if it's possible for this version to ever start with a letter? I've been maintaining this since Xcode 8.3 and every version since then including betas start with a number https://github.com/keith/Xcode.app-strings/commits/master but the problem is actually the case where they _end_ with a letter.

Here's the exception that I hit when I tried that originally:

```
Caused by: java.lang.IllegalArgumentException: Dotted version components must all be of the form \d+([a-z]+\d*)? but got 10.2.0.10p91b
```

(note I lowercased it locally because the regex is case sensitive)

It looks DottedVersion is very tied to there only being a single alpha group in each component, otherwise it looks like it would break sorting

**Comment by aragos** [comment]
I'd be fine with modifying DottedVersion to be more permissive if you feel like that's required to represent Apple's naming scheme. Both upper casing and more letters should be ok.

**Comment by keith** [comment]
I'm not sure that changing DottedVersion is viable. The implementation of it is very tied to its internal `Components` type that is very specific about the alpha requirements. I'm not sure we could reasonable support the current behavior and this new format.

**Comment by keith** [comment]
I talked to Sergio about this more offline and am taking a stab at updating DottedVersion

File: tools/osx/xcode_locator.m [hunk 1]
```diff
@@ -259,12 +270,6 @@ int main(int argc, const char * argv[]) {
         versionArg = @"";
       } else {
         versionArg = firstArg;
-        NSCharacterSet *versSet =
-            [NSCharacterSet characterSetWithCharactersInString:@"0123456789."];
-        if ([versionArg rangeOfCharacterFromSet:versSet.invertedSet].length
-            != 0) {
-          versionArg = nil;
-        }
```
Comment (keith) [inline tools/osx/xcode_locator.m]: In theory we could add all uppercase and lowercase letters here, but I don't think this validation added a lot of value anyways, so it's probably fine to remove it entirely

**Comment by keith** [comment]
@aragos I updated DottedVersion to support this use case so this is ready for another pass!

**Comment by keith** [comment]
Awesome thanks!

**Comment by DavidGoldman** [comment]
It's worth noting that even with this change Bazel may not select the proper Xcode beta if you don't specify the `--xcode_version` flag via the command line and instead rely on the active `xcode-select`'ed Xcode version: due to the version scanning [here](https://github.com/bazelbuild/bazel/blob/c6c9b58f14ec828cc050f0e6d201316bd986cf3f/tools/osx/xcode_configure.bzl#L212) which does not check the build version. It's probably best way that code to use `xcode-select -p` directly.

**Comment by keith** [comment]
+1 that's a good callout, my expectation was generally just that if you're not setting that you're not super worried about what Xcode version people are using in general, so it would probably be ok

**Pull request closed without merging**